pkg/controller/volume/snapshot/snapshot_controller.go

```diff
@@ -0,0 +1,1470 @@
+	"k8s.io/kubernetes/pkg/util/goroutinemap"
+	"k8s.io/kubernetes/pkg/util/goroutinemap/exponentialbackoff"
+	vol "k8s.io/kubernetes/pkg/volume"
+)
+
+// Number of retries when we create a VSD object for a created snapshot.
+const createSnapshotDataRetryCount = 5
+
+// Interval between retries when we create a VSD object for a provisioned snapshot.
+const createSnapshotDataInterval = 10 * time.Second
+
+// defaultSnapshotSyncPeriod is the default period for syncing volume snapshots
+// and volume snapshot datas.
+const defaultSnapshotSyncPeriod = 20 * time.Second
+
+// annBindCompleted annotation applies to snapshots. It indicates that the lifecycle
+// of the snapshot has passed through the initial setup. This information changes how
+// we interpret some observations of the state of the objects. Value of this
+// annotation does not matter.
+const annBindCompleted = "vsd.kubernetes.io/bind-completed"
+
+// This annotation is added to a VSD that has been dynamically created by
+// Kubernetes. Its value is name of volume plugin that created the snapshot data.
+// It serves both user (to show where a VSD comes from) and Kubernetes (to
+// recognize dynamically created VSDs in its decisions).
+const annDynamicallyCreatedBy = "vsd.kubernetes.io/created-by"
+
+// annBoundByController annotation applies to VSDs and Vss.  It indicates that
+// the binding (VSD->VS or VS->VSD) was installed by the controller.  The
+// absence of this annotation means the binding was done by the user (i.e.
+// pre-bound). Value of this annotation does not matter.
+const annBoundByController = "vsd.kubernetes.io/bound-by-controller"
+
+// This annotation is added to a volumeSnapshot that is supposed to create snapshot
+// Its value is name of volume plugin that is supposed to create
+// a snapshot data for this snapshot.
+const annStorageSnapshotter = "snapshot.beta.kubernetes.io/storage-snapshotter"
+
+// annSnapshotDataShouldDelete is added to a VSD that will be deleted. It serves external
+// snapshotter to delete the backend snapshot data. Value of this annotation does not matter.
+const annSnapshotDataShouldDelete = "vsd.kubernetes.io/should-delete"
+
+// ControllerParameters contains arguments for creation of a new
+// SnapshotController controller.
+type ControllerParameters struct {
+	KubeClient                 clientset.Interface
+	SyncPeriod                 time.Duration
+	VolumePlugins              []vol.VolumePlugin
+	Cloud                      cloudprovider.Interface
+	ClusterName                string
+	VolumeInformer             coreinformers.PersistentVolumeInformer
+	ClaimInformer              coreinformers.PersistentVolumeClaimInformer
+	VolumeSnapshotInformer     storageinformers.VolumeSnapshotInformer
+	VolumeSnapshotDataInformer storageinformers.VolumeSnapshotDataInformer
+	EventRecorder              record.EventRecorder
+}
+
+type SnapshotController struct {
+	vsLister           storagelisters.VolumeSnapshotLister
```

Comment:
maybe snapshotLister and snapshotDataLister?

Reply:
VolumeSnapshotLister is an auto-generated name.  The only way to change it is to rename the API object from VolumeSnapshot to Snapshot and then re-generate the Lister.

Reply:
@jingxu97 You meant to rename "vsLister" to "snapshotLister"?  Sure.